games/xworld3d/x3item.cpp

```diff
@@ -0,0 +1,144 @@
+    object_.destroy();
+}
+
+/*********************************** X3Block***********************************/
+
+/*********************************** X3Goal ***********************************/
+
+/*********************************** X3Agent **********************************/
+X3Agent::X3Agent(const X3ItemInfo& info, World& world,
+                 float speed_norm, int orientation_bins,
+                 float reaching_dist) :
+        X3Item(info, world),
+        speed_norm_(speed_norm),
+        orientation_bins_(orientation_bins),
+        reaching_dist_(reaching_dist) {
+    CHECK(info.type == X3EntityType::AGENT);
+    yaw_id_ = simulator::util::get_rand_ind(orientation_bins_);
+    set_direction();
+}
+
+void X3Agent::move_forward() {
+    float vx = speed_norm_ * dir_x_;
+    float vy = speed_norm_ * dir_y_;
+    float vz = object_.speed_z();
+    set_pose_and_speed(pose(), vx, vy, vz);
+}
+
+void X3Agent::turn_left() {
+    yaw_id_ = (yaw_id_ + 1) % orientation_bins_;
+    float vz = object_.speed_z();
+    set_speed(0.0f, 0.0f, vz);
+    set_direction();
+}
+
+void X3Agent::turn_right() {
+    yaw_id_ = (yaw_id_ - 1) % orientation_bins_ + orientation_bins_;
+    float vz = object_.speed_z();
+    set_speed(0.0f, 0.0f, vz);
+    set_direction();
+}
+
+void X3Agent::jump() {
+    if (fabs(pose().z()) < EPSILON) {
+        set_speed(0.0f, 0.0f, speed_norm_);
+    }
+}
+
+float X3Agent::reach_test(const Pose& gpose) {
+    float dx = gpose.x() - pose().x();
+    float dy = gpose.y() - pose().y();
+    float dz = gpose.z() - pose().z();
+    float d = sqrt(dx * dx + dy * dy);
+    float reaching_score = -2; // a value smaller than any possible sim value
+    if (d < reaching_dist_ && dz < 0.05) {
+        dx /= d;
+        dy /= d;
+        reaching_score = dx * dir_x_ + dy * dir_y_;
+    }
+
+    return reaching_score;
+}
+
+void X3Agent::set_direction() {
+    float rad = 2 * yaw_id_ * M_PI / orientation_bins_;
+    dir_x_ = cos(rad);
+    dir_y_ = sin(rad);
+}
+
+/********************************** X3Camera **********************************/
+X3Camera::X3Camera(World& world, int img_height, int img_width) :
+        camera_(world.new_camera_free_float(img_width, img_height, "camera")),
+        agent_(NULL) {}
+
+void X3Camera::attach_agent(X3Agent* agent) {
+    if (agent && agent_ != agent) {
+        agent_ = agent;
+    }
+}
+
+void X3Camera::update(bool debug) {
+    // TODO: to support rendering using detached camera
+    CHECK(agent_) << "camera is detached";
+    Pose p = agent_->pose();
+    if (!debug) {
+        double dir_x, dir_y;
+        agent_->get_direction(dir_x, dir_y);
+        camera_.move_and_look_at(p.x(), p.y(), p.z(),
+                                 p.x() + dir_x, p.y() + dir_y, p.z());
+    } else {
+        // bird view
+        camera_.move_and_look_at(p.x(), p.y(), 1.0f, p.x(), p.y(), p.z());
+    }
+}
+
+roboschool::RenderResult X3Camera::render(X3Agent* agent, bool is_debug) {
```

Comment:
suggest to change variable "is_debug" to a better name reflecting it corresponds to bird_view if turned on.